### ISSUE TITLE
Jsr223 Rule-Id Generation

### ISSUE THREAD

**Issue by smerschjohann** [eclipse/smarthome#3528]
In the current master branch, the generation of ruleids für jsr223 rules is broken. The following change has to be changed or reverted: https://github.com/eclipse/smarthome/commit/55e50bb55eb848710cb4c502e3b6902722ebdc5b#diff-4bf68980513df193d701ca2fdaceb9cbR172

It seems that the hashCode() cannot be used to uniquely identify the scripts. Futhermore, the "new names" do not contain the classnames of the scripts anymore. Before this change, the names had the jython/groovy etc. classname in it which helped to identify the script it belonged to.

**Comment by smerschjohann** [comment]
see also the discussion here: https://community.openhab.org/t/openhab-2-jython-support/27777/14

**Comment by kaikreuzer** [comment]
> It seems that the hashCode() cannot be used to uniquely identify the scripts

Imho it should - so if it currently does not result in a proper hash value, the implementation of the `hashCode()` method should be changed in the rules.

> do not contain the classnames of the scripts anymore

Right, this resulted in far too long IDs (and incredibly ugly labels, which weren't usable in the UI).

If people need a way to identify where the rules come from, shouldn't there simply be a way for the user to define an id himself within a rule, just like every Xtext rule has a user-defined ID as well?

**Comment by maggu2810** [comment]
Isn't it a common mistake to use the `hashCode()` method for a unique identification?

**Comment by maggu2810** [comment]
Isn't it a common mistake to use the `hashCode()` method for a unique identification?

There are several articles about the topic e.g. part 2 "HashCode collisions" of http://eclipsesource.com/blogs/2012/09/04/the-3-things-you-should-know-about-hashcode/

**Comment by kaikreuzer** [comment]
Thanks, you are obviously right that this was indeed a bad idea.

Having another look at it, I have now adapted the code to follow the idea of the Rule Engine in general: If there is no user defined id provided, it is simply left to the Rule Engine itself to determine a unique id for the rule. Additionally, I have adapted the ids of the modules as well, since they can simply be continuously numbered (as they only have to be unique within a rule).

See https://github.com/eclipse/smarthome/pull/3534.
